agent/http.go

```diff
@@ -104,6 +106,24 @@ var endpoints map[string]unboundEndpoint
 // An empty slice means an endpoint handles OPTIONS requests and MethodNotFound errors itself.
 var allowedMethods map[string][]string = make(map[string][]string)
 
+// contextKey is a value for use with context.WithValue. It's used as
+// a pointer so it fits in an interface{} without allocation.
+
+// TODO: (spatel) Is there a way to continue using a contextKey without a circular import
+//                when referenced by consul/agent/server.go?
+
+// type contextKey struct {
+// 	name string
+// }
+
+// func (k *contextKey) String() string { return "agent/http context value " + k.name }
+
+// // RemoteAddrContextKey is a context key. It can be used in
+// // HTTP handlers with Context.Value to access the remote
+// // address the connection arrived on.
+// // The associated value will be of type net.Addr.
+// var RemoteAddrContextKey = &contextKey{"remote-addr"}
+
```

Comment:
The context key is currently a string `"remote-addr"`. Golang docs recommend against used a string to prevent name collisions. However, using a type results in a circular dep (See comment on lines 112-113). Anything obvious I'm missing here or will a string suffice for now?

Reply:
I'd recommend creating helper methods in a package that both `agent` and `agent/consul` can import without cycles. Maybe even in `agent/consul` itself:

```Go
package consul

type contextKeyRemoteAddr struct{}

func ContextWithRemoteAddr(ctx context.Context, addr net.Addr) context.Context {
	return context.WithValue(ctx, contextKeyRemoteAddr{}, addr)
}

func RemoteAddrFromContext(ctx context.Context) (net.Addr, bool) {
	v, ok := ctx.Value(contextKeyRemoteAddr{})
	if !ok {
		return nil, false
	}
	return v.(net.Addr), true
}
```

```Go
func withRemoteAddrHandler(next http.Handler) http.Handler {
	return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
		addrPort, err := netip.ParseAddrPort(req.RemoteAddr)
		if err == nil {
			remoteAddr := net.TCPAddrFromAddrPort(addrPort)
			ctx := consul.ContextWithRemoteAddr(req.Context(), remoteAddr)
			req = req.WithContext(ctx)
		}
		next.ServeHTTP(resp, req)
	})
}
```

Reply:
I like it! Done.